presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -1397,6 +1408,18 @@ private ThriftMetastoreClient createMetastoreClient()
         return clientProvider.createMetastoreClient();
     }
 
+    private ThriftMetastoreClient createMetastoreClient(HiveContext context)
+            throws TException
+    {
+        ThriftMetastoreClient client = createMetastoreClient();
+        if (!impersonationEnabled) {
+            return client;
+        }
+
+        client.setUGI(context.getUsername());
```

Comment:
if this throws, you should call `client.close()`

Reply:
We have the same problem in `StaticMetastoreLocator`. Maybe create a helper method
```java
static void setMetastoreUserOrClose(ThriftMetastoreClient client, String user)
{
    try {
        client.setUGI(user);
    }
    catch (Throwable t) {
        // close client and suppress any error from close
        try (Closeable ignored = client) {
            throw t;
        }
    }
}